src/basic/user-util.c

```diff
@@ -238,14 +238,21 @@ int get_user_creds(
         }
 
         if (home) {
-                if (FLAGS_SET(flags, USER_CREDS_CLEAN) && empty_or_root(p->pw_dir))
-                        *home = NULL;
+                if (FLAGS_SET(flags, USER_CREDS_CLEAN) &&
+                    (empty_or_root(p->pw_dir) ||
+                     !path_is_valid(p->pw_dir) ||
+                     !path_is_absolute(p->pw_dir)))
+                    *home = NULL; /* Note: we don't insist on normalized paths, since there are setups that have /./ in the path */
                 else
                         *home = p->pw_dir;
         }
 
         if (shell) {
-                if (FLAGS_SET(flags, USER_CREDS_CLEAN) && (isempty(p->pw_shell) || is_nologin_shell(p->pw_shell)))
+                if (FLAGS_SET(flags, USER_CREDS_CLEAN) &&
+                    (isempty(p->pw_shell) ||
+                     !path_is_valid(p->pw_dir) ||
+                     !path_is_absolute(p->pw_shell) ||
```

Comment:
Is absoluteness really necessary? We moved towards not requiring full paths in ExecStart, and I don't see why we shouldn't relax things here in the same way.

Reply:
ExecStart= is something where we can define the execution environment, and hence can clearly define what $PATH means in that case. Which makes the whole thing acceptable I figure, since it's clearly defined what happens.

However, for the user db this is different: it has many consumers, hence the only thing to make this all somewhat generally supportable is by insisting on an absolute paths, since only then it is clear what this means...

Note that this bit here is pretty defensive: it doesn't refuse anything, it doesn't fail. it just resets the field to NULL, which effectively means /bin/sh is used after all, and that's a good thing I think. i.e. it's not a question of allow/deny but a question of "let's revert to defaults if something is fishy"

Reply:
I think we *could* apply the same logic, and search for the executable using our built-in path. The obvious advantage is that it allows specifying the shell without bothering to figure out the exact path (sbin vs. bin, etc). But it's probably OK to make this strict, since that's how shells were so far always specified.